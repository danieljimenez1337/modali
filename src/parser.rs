use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum Action {
    SubAction {
        key: String,
        description: String,
        sub_actions: Vec<Action>,
    },

    KeyAction {
        key: String,
        description: String,
        command: String,
    },
}

impl Action {
    pub fn to_tree_node(&self) -> WhichTreeNode {
        match self {
            Action::SubAction {
                key,
                description,
                sub_actions,
            } => {
                let label = format!("{key}: {description}");
                let children = sub_actions
                    .iter()
                    .map(|action| action.to_tree_node())
                    .collect();
                WhichTreeNode {
                    key: parse_char(key),
                    label,
                    kind: WhichTreeKind::Children(children),
                }
            }
            Action::KeyAction {
                key,
                description,
                command,
            } => {
                let label = format!("{key}: {description}");
                WhichTreeNode {
                    key: parse_char(key),
                    label,
                    kind: WhichTreeKind::Command(command.clone()),
                }
            }
        }
    }
}

pub fn actions_to_tree(actions: &[Action]) -> Vec<WhichTreeNode> {
    actions.iter().map(|a| a.to_tree_node()).collect()
}

// pub fn search_which_tree(tree: &[WhichTreeNode], s: String) -> WhichTreeNode

fn parse_char(s: &str) -> char {
    if s.chars().count() != 1 {
        panic!("This key is not a single char")
    }
    s.chars().next().unwrap()
}

pub struct WhichTreeNode {
    key: char,
    label: String,
    kind: WhichTreeKind,
}

pub enum WhichTreeKind {
    Command(String),
    Children(Vec<WhichTreeNode>),
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    #[test]
    fn test_load_config_file() {
        let contents = fs::read_to_string("bindings.json").unwrap();
        let config: Vec<Action> = serde_json::from_str(&contents).unwrap();
        let json = serde_json::to_string_pretty(&config).unwrap();
        insta::assert_snapshot!(json)
    }
}
