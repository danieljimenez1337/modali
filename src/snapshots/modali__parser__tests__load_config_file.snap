---
source: src/parser.rs
expression: json
---
[
  {
    "type": "KeyAction",
    "key": "r",
    "description": "Open Rofi",
    "command": "rofi -show drun"
  },
  {
    "type": "KeyAction",
    "key": "g",
    "description": "Go to Google",
    "command": "xdg-open https://google.com"
  },
  {
    "type": "SubAction",
    "key": "f",
    "description": "Firefox options...",
    "sub_actions": [
      {
        "type": "KeyAction",
        "key": "f",
        "description": "Launch Firefox",
        "command": "firefox"
      },
      {
        "type": "KeyAction",
        "key": "e",
        "description": "Edit this config",
        "command": "xdg-open bindings.json"
      }
    ]
  }
]
